window-icon-updater/icon-sender

```
@@ -82,7 +84,133 @@ class IconRetriever(object):
+                    # A menu, pop-up or similar icon-less window
+                    raise NoIconError()
+                icon_pixmap_geometry = self.conn.core.GetGeometry(
+                    atoms[3]
+                ).reply()
+                if (
+                        # Only 1bit/pixel & 24bit/pixel icons are tested
+                        not icon_pixmap_geometry.depth in [1, 24]
+                        or icon_pixmap_geometry.width > ICON_MAX_SIZE
+                        or icon_pixmap_geometry.height > ICON_MAX_SIZE
+                ):
+                    raise NoIconError()
+                image = self.conn.core.GetImage(
+                    xproto.ImageFormat.ZPixmap,
+                    atoms[3],
+                    0,
+                    0,
+                    icon_pixmap_geometry.width,
+                    icon_pixmap_geometry.height,
+                    0xFFFFFFFF
+                ).reply()
+                icon_pixmap_data = image.data.raw
+                if not flags & IconMaskHint:
+                    # We have an icon without transparency (mask)
+                    icon_mask_geometry = None
+                    icon_mask_data = None
+                else:
+                    icon_mask_geometry = self.conn.core.GetGeometry(
+                        atoms[7]
+                    ).reply()
+                    image = self.conn.core.GetImage(
+                        xproto.ImageFormat.ZPixmap,
+                        atoms[7],
+                        0,
+                        0,
+                        icon_mask_geometry.width,
+                        icon_mask_geometry.height,
+                        0xFFFFFFFF
+                    ).reply()
+                    icon_mask_data = image.data.raw
+            except (xproto.BadWindow, xproto.WindowError, xproto.AccessError):
+                raise NoIconError
+            except Exception:   # pylint: disable=bare-except
+                raise NoIconError
```

What exactly falls into this? Silently ignoring all other errors is a bad idea. At the very least add a log message, but you can also simply drop this one, to not continue in case of fatal error (like connection to X server lost).

---

> But I tried the code without it and it fails

Then not ignoring that exception is even more important to find out what is wrong. Maybe `traceback.print_exc()` will help? And/or printing `repr(e)` or os.

> But it gives no clue (It only catch `NoIconError` errors).

`NoIconError` should _not_ be handled here, it should be handled by the caller.

---

Just remembered how to reproduce the bug. It happens if you install `xlogo` in Fedora without `xorg-x11-xbitmaps` & `xorg-x11-fonts-75dpi` packages. And they you try to run `xlogo`. In that case, it crashes it (the crash happens within the try/except.

xlogo prints "Warning: Cannot convert string "xlogo32" to type Pixmap"

While this is a failure of Fedora packaging of setting the dependencies, still the code should be resilient.

---

Found it. Additional error to catch is `xcffib.xproto.DrawableError`. I will amend the PR in a minute.